From: Lorenzo Stoakes <lorenzo.stoakes@oracle.com>
Subject: mm/mseal: add comment explaining why we disallow gaps on mseal()
Date: Fri, 25 Jul 2025 19:23:50 +0100

This explains the semantics clearly, the 'why' of the situation.

Link: https://lkml.kernel.org/r/d85b3d55-09dc-43ba-8204-b48267a96751@lucifer.local
Signed-off-by: Lorenzo Stoakes <lorenzo.stoakes@oracle.com>
Cc: David Hildenbrand <david@redhat.com>
Cc: Jann Horn <jannh@google.com>
Cc: Jeff Xu <jeffxu@chromium.org>
Cc: Kees Cook <kees@kernel.org>
Cc: Liam R. Howlett <Liam.Howlett@oracle.com>
Cc: Pedro Falcato <pfalcato@suse.de>
Cc: Vlastimil Babka <vbabka@suse.cz>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/mseal.c |   12 ++++++++++++
 1 file changed, 12 insertions(+)

--- a/mm/mseal.c~mm-mseal-simplify-and-rename-vma-gap-check-fix
+++ a/mm/mseal.c
@@ -38,6 +38,18 @@ out:
 }
 
 /*
+ * mseal() disallows an input range which contain unmapped ranges (VMA holes).
+ *
+ * It disallows unmapped regions from start to end whether they exist at the
+ * start, in the middle, or at the end of the range, or any combination thereof.
+ *
+ * This is because after sealng a range, there's nothing to stop memory mapping
+ * of ranges in the remaining gaps later, meaning that the user might then
+ * wrongly consider the entirety of the mseal()'d range to be sealed when it
+ * in fact isn't.
+ */
+
+/*
  * Does the [start, end) range contain any unmapped memory?
  *
  * We ensure that:
_
